keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1267 @@
+we extend the Pod delete API to pass the reason for pod termination. Such
+failures can typically be associated with some cluster-management actions and
+should be retried.
+
+Additionally, we review all the currently existing places were the delete API
+is invoked to delete pods in order to modify the code to pass a meaningful
+reason based on the invocation context.
+
+We use the main job controller main loop to detect and categorize the pod failures
+with respect to the configuration. For each failed pod, one of the following
+actions is applied:
+- terminate the job (`non-retriable` failure),
+- ignore the failure (`retriable` failure) - restart the pod and do not increment
+  the counter for `backoffLimit`,
+- increment the `backoffLimit` counter and restart the pod if the limit is not
+  reached (current behaviour).
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a machine learning researcher, I run jobs comprising thousands
+of long-running pods on a cluster comprising thousands of nodes. The jobs often
+run at night or over weekend without any human monitoring. In order to account
+for random infrastructure failures we define `.backoffLimit: 6` for the job.
+However, a signifficant portion of the failures happen due to bugs in code.
+Moreover, the failures may happen late during the program execution time. In
+such case, restarting such a pod results in wasting a lot of computational time.
+
+We would like to be able to automatically detect and terminate the jobs which are
+failing due to the bugs in code of the executable, so that the computation resources
+can be saved.
+
+Occasionally, our executable fails, but it can be safely restarted with a good
+chance of succeeding the next time. In such known retriable situations our
+executable exits with a dedicated exit code in the 40-50 range. All remaining
+exit codes indicate a software bug and should result in an early job termination.
+
+The following Job configuration could be a good starting point to satisfy
+my needs:
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  template:
+    spec:
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./program"]
+  backoffLimit: 3
+  backoffPolicy:
+    rules:
+    - containerExitCodeNotIn: 40-50
```

Comment:
I think this should be something like:

```yaml
containerExitCode:
  notIn: 40-50
```

or

```yaml
containerExitCode:
  value: 40-50
  operator: NotIn
```

Reply:
I was thinking about it, but I feel that another level of nesting is over-complicating the API.

Reply:
Is it really a good idea to specify two numbers in a single string `40-59`? This feels like a YAML parsing trap, and creates trouble for automation (e.g. I can't easily `jq '.spec.backoffPolicy.rules[0].containerExitCodeNotIn.max = 59`).

Reply:
Good points, I will define ranges (min, max) and extract the "in" and "notIn" to separate levels.
See the proposal below (I will update the KEP accordingly):

```
  backoffPolicy:
    rules:
    - action: Terminate
      containerName: main-job-container
      onContainerExitCode:
        inRange:
        - min: 1
          max: 12
        - min: 100
          max: 127
    - action: Ignore
      onPodStatusReason:
        in:
        - Preemption
        - NodePressureEviction
```

Reply:
So each exit code filter is an `IntOrIntRange`?

Reply:
Also note the possible off-by-one ambiguity. Does `{min: 1, max: 12}` include 12 or not?

Reply:
@SOF3 how about:
- introducing "inRanges" along with "in" (to avoid "IntOrIntRange")
- rename the min, max -> minInclusive and maxInclusive. 

Then, the example could be formulated as:
```
  backoffPolicy:
    rules:
    - action: Terminate
      containerName: main-job-container
      onContainerExitCode:
        inRanges:
        - minInclusive: 1
          maxInclusive: 12
        - minInclusive: 100
          maxInclusive: 127
    - action: Ignore
      onPodStatusReason:
        in:
        - Preemption
        - NodePressureEviction
```

Reply:
Update, in the latest API proposal this can be expressed as follows (also notice the change too lookup for PodCondition rather than Pod StatusReason):

```yaml
 backoffPolicy:
   rules:
   - action: Terminate
     requirements:
     - onExitCodes:
         containerName: main-job-container
         inRange:
         - minInclusive: 1
           maxInclusive: 10
         - minInclusive: 100
           maxInclusive: 127
   - action: Ignore
     requirements:
     - onPodCondition:
       - key: Status
         operator: In
         values:
         - True
       - key: Type
         operator: In
         values:
         - Preempted
         - Evicted
```